psh/date/date.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Phoenix-RTOS
+ *
+ * date - print or set the system date and time
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+enum mode { SKIP,
+	HELP,
+	PRINT,
+	WRITE };
+
+static void psh_date_info(void)
+{
+	printf("print/set the system date and time");
+}
+
+
+static void psh_date_help(const char *prog)
+{
+	printf("Usage: %s [-h] [-s EPOCH] [+FORMAT]\n", prog);
+	printf("  -h:  shows this help message\n");
+	printf("  -s:  set system time described by EPOCH (POSIX time format)\n");
+	printf("  FORMAT: string with POSIX date formatting characters\n");
+	printf("NOTE: FORMAT string not supported with '-s' option\n");
+}
+
+
+static int psh_date_print(char *format)
+{
+	time_t t;
+	struct tm *tinfo;
+	char defformat[] = "+%a, %d %b %y %H:%M:%S";
+	char buf[64];
+
+	if (format == NULL) {
+		format = defformat;
+	}
+	else if (format[0] != '+') {
+		fprintf(stderr, "date: invalid date '%s'\n", format);
+		return 1;
+	}
+
+	time(&t);
+	tinfo = localtime(&t);
+
+	/* passing `&format[1]` for omission of '+' at the beginning */
+	if (strftime(buf, 64, &format[1], tinfo) == 0) {
+		fprintf(stderr, "date: '%s' expansion too long\n", format);
+		return 1;
+	}
+	printf("%s\n", buf);
+
+	return 0;
+}
+
+
+static int psh_date_set(char *timestring, const char *format)
+{
+	struct timeval tv;
+	struct timezone tz = { .tz_dsttime = 0, .tz_minuteswest = 0 };
+	char *end;
+
+	/* FIXME: no strptime() implementation in libphoenix */
+	if (format != NULL) {
```

Comment:
still, I would appreciate if `date -s @12345678` would work

Reply:
added proper `date -d @123456` parsing ability. Also added tests for that